### ISSUE TITLE
possible DOSXYZ bug with collimator angle (phicol)

### ISSUE THREAD

**Issue by gwu-reflexionmedical** [nrc-cnrc/EGSnrc#141]
I used a same phase space source at multiple positions (top, left, bottom and right), the results are normal from top, bottom and right, but one the right, it seems like DOSXYZ automatically and mistakenly rotate the phase space source. See the attached picture:
[https://www.dropbox.com/s/o4eiygbe81vt0cd/Selection_041.png?dl=0](https://www.dropbox.com/s/o4eiygbe81vt0cd/Selection_041.png?dl=0)
(The left source is actual shown in the right of the image, need to teach matlab how to show the axis at some point )

I tried this on three different settings with same **Phicol** value : 
(1) four individual simulation, with only (theta, phi) differs between input files top: (180, 0), right: (90, 0),  bottom: (0,0), left: (90,180).
(2) One single simulation with source = 8, (One phase space source at multiple position), the same theta-phi pairs are used, each was assigned a probability of 0.25. 
(3) slightly off theta-phi pairs, top: (179.9, 0.1), right: (89.9, 0.1),  bottom: (0.1, 0.1), left: (89.9, 179.9).

All results are the same, something is not quite right in the code. The input file is here:
[https://www.dropbox.com/s/874ihhnldk5e63p/edge_rot.egsinp?dl=0](https://www.dropbox.com/s/874ihhnldk5e63p/edge_rot.egsinp?dl=0)
I used the latest EGSnrc in Ubuntu 15.10 with its default gcc, g++, gfortran compilers. 

BTW, the description for *Phicol* is quite confusing in the dosxyznrc manual. In some places, it suggests that *phicol = 180* when passing BEAM simulated phase space to DOSXYZ. While in some places, it suggests to set phicol only when (1) passing BEAM result to DOSXYZ and (2) theta = 180 degree. (My understanding is that phicol should be 180 degree as long as the phase space source is from BEAM, no matter what theta angle the source locates). Also, the description that " phicol is determined at theta = phi = 0" which is not correct, one still need to set the phicol to 180 if he uses the BEAM simulated phase space source.

**Comment by gwu-reflexionmedical** [comment]
To further test this, I run a DOSXYZ simulation with one phase space source (source = 8) at multiple position. I set the source to be evenly distributed over the space (Like a ring), I am expecting to see a dose "ring" in the phantom, however, I only a half ring due to the "automatic" **Phicol** shift (Radiation beams that should form the top ring did actually amplify the bottom ring). Please check the picture

[https://www.dropbox.com/s/6atg4e3lzs50d08/Selection_042.png?dl=0](url)

**Comment by rtownson** [comment]
There have been a number of papers published on coordinate transformations for DOSXYZnrc - it has been this way for a long time. Even if the coordinate system is not intuitive, I don't think this is something we could change at this stage. In the future, maybe we will look at including an optional alternative coordinate system using DICOM patient coordinates?

**Comment by rtownson** [comment]
The dropbox links do not work for me.

**Comment by gwu-reflexionmedical** [comment]
Thanks @rtownson , I fixed the url thing.

I think what happened is that the actual code does not match the description on the coordinate system in the **DOSXZYnrc** manual. The manual suggests that **0 < theta < pi** and  **0 < phi < 2*pi**, however, the **phicol** automatically shift when I follow this setting (source rotates on one side). I did some additional tests that show if I use **0 < theta < 2*pi** or **-pi < theta < pi** with **phi** fixed at 0, the **phicol** will be ok (not be shifted). This inconsistent between the code and the manual is quite confusing and could be improved. 

[Here is the new simulation result, which shows a complete ring (source did not rotate on one side)](https://www.dropbox.com/s/hiaq1ddtzaly4ud/Selection_043.png?dl=0)

**Comment by marenaud** [comment]
I don't think phicol itself changes. When the manual says that phicol is determined for theta = phi = 0, it just means that the phicol rotation (basically a rotation about the Z axis if I'm not mistaken) is applied to the particles before the theta and phi rotations, and the particles are initially on a plane normal to the Z axis. The confusing part is that the name phicol implies that it should be the physical collimator angle when it really isn't. If you modelled a non coplanar trajectory with a physical collimator angle of 45 degrees the whole way around, phicol wouldn't be 45 across the entire trajectory.

The transformations to go from the gantry/couch/collimator coordinate system to DOSXYZ are given in a couple papers as Reid Townson noted,
Thebaut, J., and S. Zavgorodni. "Coordinate transformations for BEAM/EGSnrc Monte Carlo dose calculations of non-coplanar fields received from a DICOM-compliant treatment planning system." Physics in medicine and biology 51.23 (2006): N441.

Zhan, Lixin, Runqing Jiang, and Ernest K. Osei. "Beam coordinate transformations from DICOM to DOSXYZnrc." Physics in medicine and biology 57.24 (2012): N513.

etc etc

**Comment by rtownson** [comment]
Thanks @marenaud, I re-opened this to allow for discussion on whether/how pirs794 should be clarified.

**Comment by gwu-reflexionmedical** [comment]
Thanks @marenaud . I agree that phicol itself should not change. What I found is if I use the convention for theta and phi, the DOSXYZ will rotate my source, which is equivalent to that phicol is shifted by 180.

So it could be that theta-phi is not implemented as what the manual suggests, or that a bug in the code that shifted the phicol when the user do not want it to be shifted.

**Comment by marenaud** [comment]
If you look at the code, everything is implemented as advertised in the manual (or seems to be at first glance), it's just that the coordinate system in DOSXYZ is not the same as the gantry/couch/collimator coordinate system so a bit of work has to be done to make them match. The manual itself doesn't make any false claims, and doing the transformations outlined in those papers will net you the results you expect, assuming they didn't make a mistake in deriving the transformations ;)

The particle rotations are just a series of 3 rotations applied sequentially on the initial particle position. Since sin and cos functions are defined for any angle range, you'll still get sensible results if you redefine the ranges of theta and phi, but then the manual's conventions will no longer apply. Phicol doesn't ever get modified during the simulation, the rotation angles are precalculated for each field beforehand in srcxyznrc (or on the fly for some sources, but the rotation matrix is the same)

```
"Rotation matrix needed during sampling routine
r_11(1) =  costheta*cosphi*coscol + sinphi*sincol;
r_12(1) = -costheta*cosphi*sincol + sinphi*coscol;
r_13(1) = -sintheta*cosphi                       ;
r_21(1) =  costheta*sinphi*coscol - cosphi*sincol;
r_22(1) = -costheta*sinphi*sincol - cosphi*coscol;
r_23(1) = -sintheta*sinphi                       ;
r_31(1) = -sintheta*coscol                       ;
r_32(1) =  sintheta*sincol                       ;
r_33(1) = -costheta                              ;
```

**Comment by gwu-reflexionmedical** [comment]
@marenaud, thank you very much for this detailed reply.  I understand that the coordinate system is different between what is in the imaging systems (recorded in DICOM) and the one in DOSXYZ. What I confused about is that if I do a 360 degree scan followed by the manual, the source orientation will be different between the left and right space. 

For example, the manual suggests **0 < theta < pi** and **0 < phi < 2*pi**, so if I want to do a 360 degree scan, I need to use two **phi** values, **phi=0 degree* for the source in the positive x direction, and **phi=180 degree* for source in the negative x direction. However, the simulation shows (90, 0, 180) (theta, phi, phicol) and (90, 180, 180) (theta, phi, phicol) have two different source orientation, even though the **phicol** value in both simulations are the same... 

If I do (90, 0, 180) and (270, 0, 180) simulation, everything will be correct and the source will not be rotated. So I am wondering whether I should use  **0 < theta < 2*pi** or use what suggested in the manual, which is **0 < theta < pi**. (This also requires to use **-pi < phi < pi**)

**Comment by gwu-reflexionmedical** [comment]
@marenaud . I think you are right, one need to re-define the range for **theta** and **phi** in order to get DOSXYZ working. I think the property of `cos()` function and the way it is used requires DOSXYZ user to use a different range for phi and theta. 

It would be great to have this info in the manual, or delete the part that suggests **0 < theta < pi and 0 < phi < 2*pi** in the manual

**Comment by marenaud** [comment]
The original range in the manual seems fine, it basically just mimics the spherical coordinate system (without phicol). I don't think you're out of the woods if you redefine the ranges of theta and phi, your examples are "working" with the new definition because of the test cases you're using.

If you tried delivering an asymmetric field with no symmetry axes about the origin like a 1x1 field with x = (4, 5), y = (4, 5) and you aligned it with the +y and -y axes with your angle definitions, you'd get something like theta, phi, phicol =  (90, 90, 180) and (270, 90, 180). I think you would still see that with the same value of phicol, the beam will have a different orientation. (I might be wrong about that, it's hard to visualise without doing the actual rotations).

In other words, you shouldn't expect that a constant value of phicol will give you a constant collimator orientation no matter what (theta, phi) are, because of the way the rotations are performed.

**Comment by gwu-reflexionmedical** [comment]
@marenaud Do you mean the coordinate transformation should be done strictly followed the way suggested in the cited papers, and my solution (which is extending the theta from 0 to 2pi) might not work in other test case? I'm a little surprised that the transformation has not being taken care of by DOSXYZ.

Your answers are very helpful, it would be nice to have that in the manual so that beginners like me would understand this topic better. Thank you !

**Comment by gwu-reflexionmedical** [comment]
@marenaud You also brought an interesting point, whether or not should DOSXYZ include **phicol pair** in the GUI for phase space source at multiple positions (source = 8). As you mentioned, the beam orientation might changed at different theta and phi even though phicol is the same. Therefore, if users, like me for example, want the beam orientation be fixed, then the **phicol** need to be changed to offset the changes in beam orientation at different positions.

Currently, the DOSXYZ only allows fixed **phicol** with varying **theta-phi pairs** for phase space source at multiple positions, which could cause problems

**Comment by marenaud** [comment]
Without having gone through a counterexample explicitly, I think that your solution won't generalise to all theta, phi, phicol possibilities (ie it won't do what you want for all angles that you can throw at it). But I'm not 100% sure so don't take my word for it.

Source 20 allows for varying phicol with a phase space file, so that should work out for you.
